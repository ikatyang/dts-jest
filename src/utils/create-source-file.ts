import * as _ts from 'typescript';

export const create_source_file = (
  filename: string,
  source: string,
  ts: typeof _ts,
) => ts.createSourceFile(filename, source, ts.ScriptTarget.Latest, true);
