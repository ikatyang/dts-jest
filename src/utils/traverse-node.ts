import * as ts from 'typescript';

export const traverse_node = (
  node: ts.Node,
  callback: (node: ts.Node) => void,
) => {
  if (node.kind !== ts.SyntaxKind.SourceFile) {
    callback(node);
  }
  ts.forEachChild(node, child => traverse_node(child, callback));
};
