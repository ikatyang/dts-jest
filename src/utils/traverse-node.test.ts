import * as ts from 'typescript';
import { create_source_file } from './create-source-file';
import { traverse_node } from './traverse-node';

const source = `

Math.max(1, 2, 3);

function abc() {
  console.log('something');
}

`;

it('should work correctly', () => {
  const source_file = create_source_file('something.ts', source, ts);

  const records: any[] = [];

  traverse_node(
    source_file,
    node => {
      records.push({
        kind: ts.SyntaxKind[node.kind],
        text: node.getText(),
      });
    },
    ts,
  );

  expect(records).toMatchSnapshot();
});
