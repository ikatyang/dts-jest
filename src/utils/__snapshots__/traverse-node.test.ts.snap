// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work correctly 1`] = `
Array [
  Object {
    "kind": "ExpressionStatement",
    "text": "Math.max(1, 2, 3);",
  },
  Object {
    "kind": "CallExpression",
    "text": "Math.max(1, 2, 3)",
  },
  Object {
    "kind": "PropertyAccessExpression",
    "text": "Math.max",
  },
  Object {
    "kind": "Identifier",
    "text": "Math",
  },
  Object {
    "kind": "Identifier",
    "text": "max",
  },
  Object {
    "kind": "FirstLiteralToken",
    "text": "1",
  },
  Object {
    "kind": "FirstLiteralToken",
    "text": "2",
  },
  Object {
    "kind": "FirstLiteralToken",
    "text": "3",
  },
  Object {
    "kind": "FunctionDeclaration",
    "text": "function abc() {
  console.log('something');
}",
  },
  Object {
    "kind": "Identifier",
    "text": "abc",
  },
  Object {
    "kind": "Block",
    "text": "{
  console.log('something');
}",
  },
  Object {
    "kind": "ExpressionStatement",
    "text": "console.log('something');",
  },
  Object {
    "kind": "CallExpression",
    "text": "console.log('something')",
  },
  Object {
    "kind": "PropertyAccessExpression",
    "text": "console.log",
  },
  Object {
    "kind": "Identifier",
    "text": "console",
  },
  Object {
    "kind": "Identifier",
    "text": "log",
  },
  Object {
    "kind": "StringLiteral",
    "text": "'something'",
  },
  Object {
    "kind": "EndOfFileToken",
    "text": "",
  },
]
`;
